File: ml-benchmark/synduce/constraints/sortedlist/max.f
Inductive List = elt Int | cons {Int, List};

is_sorted =
  let aux = fix (
    \f: Int -> List -> Bool. \pre: Int. \xs: List.
    match xs with
      elt x -> >= pre x
    | cons {h, t} -> and (>= pre h) (f h t)
    end
  ) in \xs: List.
  match xs with
    elt x -> true
  | cons {h, t} -> aux h t
  end;

max = \a: Int. \b: Int. if < a b then b else a;

maximum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt w -> w
  | cons {h, t} -> max h (f t)
  end
);

target = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    elt w -> xs
  | cons {h, _} -> xs /*Avoid recursions*/
  end
);

main = \xs: List. if is_sorted xs then maximum (target xs) else 0;

File: ml-benchmark/synduce/constraints/sortedlist/max.ml
type t = Elt of int | Cons of int * t

let is_sorted xs =
  let rec aux pre xs =
    match xs with
    | Elt x -> pre <= x
    | Cons (h, t) -> pre <= h && aux h t
  in
  match xs with
  | Elt x -> true
  | Cons (h, t) -> aux h t

let max a b =
  if a < b then b else a

let rec maximum xs =
  match xs with
  | Elt w -> w
  | Cons (h, t) -> max h (maximum t)

let target xs =
  match xs with
  | Elt w -> xs
  | Cons (h, _) -> xs

let program xs =
  if is_sorted xs then maximum (target xs) else 0